### PR TITLE
Update for symfony/process v4

### PR DESCRIPTION
**Pull request by Macavity**
Updated all usages of ProcessBuilder to use Process as the builder class was removed in version 4.0.0 of symfony/process. Sadly this makes it incompatible with older versions.

### CODE REVIEW

**Comment by elchristo** [comment]
Are there other breaking changes beside the removed ProcessBuilder class (according to the official [changelog](https://github.com/symfony/process/blob/master/CHANGELOG.md) deprecated since version 3.4.0) which require to drop version 3.* support ?

Probably it could be not too bad to proceed step by step (see PR #866) and keep 3.* version support at least temporarily.

**Comment by james-nesbitt** [comment]
Could we build a ProcessBuilder to wrap Process as a legacy handler?

File: composer.json [hunk 1]
```diff
@@ -50,7 +50,7 @@
     },
     "extra": {
         "branch-alias": {
-            "dev-master": "1.4-dev"
+            "dev-symfony-process-4": "1.5-dev"
```
Comment (stof) [inline composer.json:53]: that should not be changed at all. You don't need to alias your feature branch globally (and it makes your PR unmergeable btw)

File: src/Assetic/Filter/BaseProcessFilter.php [hunk 2]
```diff
@@ -35,11 +35,11 @@ public function setTimeout($timeout)
      *
      * @param array $arguments An optional array of arguments
      *
-     * @return ProcessBuilder A new process builder
+     * @return Process A new process builder
```
Comment (stof) [inline src/Assetic/Filter/BaseProcessFilter.php:38]: that's a BC break, so cannot be done in minor versions of Assetic.

File: src/Assetic/Filter/BaseProcessFilter.php [hunk 3]
```diff
@@ -48,10 +48,14 @@ protected function createProcessBuilder(array $arguments = array())
         return $pb;
     }
 
-    protected function mergeEnv(ProcessBuilder $pb)
+    protected function mergeEnv(Process $process)
```
Comment (stof) [inline src/Assetic/Filter/BaseProcessFilter.php:51]: The Process component allows inheriting the env on its own now (which was not the case in 2.1, which is why this logic exists)

File: src/Assetic/Filter/CleanCssFilter.php [hunk 4]
```diff
@@ -237,96 +259,96 @@ public function filterLoad(AssetInterface $asset)
...
+            : array($this->cleanCssBin);
 
         if ($this->keepLineBreaks) {
-            $pb->add('--keep-line-breaks');
+            array_push($commandline, '--keep-line-breaks');
         }
 
         if ($this->compatibility) {
-            $pb->add('--compatibility ' .$this->compatibility);
+            array_push($commandline, '--compatibility ' . $this->compatibility);
```
Comment (stof) [inline src/Assetic/Filter/CleanCssFilter.php:275]: adding the option and its value separated by a space as a single element is likely to be broken after escaping (but I think the old code was buggy already)

File: src/Assetic/Filter/CoffeeScriptFilter.php [hunk 5]
```diff
@@ -51,22 +52,22 @@ public function filterLoad(AssetInterface $asset)
...
         file_put_contents($input, $asset->getContent());
 
-        $pb = $this->createProcessBuilder($this->nodeBin
+        $commandline =$this->nodeBin
             ? array($this->nodeBin, $this->coffeeBin)
-            : array($this->coffeeBin));
+            : array($this->coffeeBin);
 
-        $pb->add('-cp');
+        array_push($commandline, '-cp');
```
Comment (stof) [inline src/Assetic/Filter/CoffeeScriptFilter.php:59]: using `$commandline[] = '-cp';` would be more readable than lots of `array_push` IMO

File: composer.json [hunk 6]
```diff
@@ -14,7 +14,7 @@
     ],
     "require": {
         "php": ">=5.3.1",
-        "symfony/process": "~2.1|~3.0"
+        "symfony/process": "~4"
```
Comment (stof) [inline composer.json:17]: this should at least allow 3.4 too, as it also has the array syntax for the Process class

File: .travis.yml [hunk 7]
```diff
@@ -14,23 +20,13 @@ cache:
...
 
 matrix:
     allow_failures:
         - php: hhvm
         - php: 7.1
-    include:
-        - php: 5.3
-          env: COMPOSER_FLAGS='--prefer-lowest --prefer-stable' SYMFONY_DEPRECATIONS_HELPER=weak
-        - php: 5.6
-          env: DEPENDENCIES=dev
```
Comment (stof) [inline .travis.yml:33]: testing lowest deps and dev deps must be kept (we could change the PHP version to a higher version for dev deps, and the lowest one should stay on the lowest supported PHP version)

**Comment by stof** [comment]
all the coding standard changes should not be included in this PR, to make the diff easier to review.

**Comment by sweiguny** [comment]
are there any updates on this topic?

**Comment by glensc** [comment]
this is too much breaking...

how about https://github.com/kriswallsmith/assetic/pull/866? or actual code changes needed for 4.x?

File: src/Assetic/Filter/BaseProcessFilter.php (hunk 2 above)
Comment (Macavity) [reply src/Assetic/Filter/BaseProcessFilter.php:38]: Agreed, but it's not my decision how you tag this afterwards.

File: src/Assetic/Filter/CoffeeScriptFilter.php (hunk 5 above)
Comment (Macavity) [reply src/Assetic/Filter/CoffeeScriptFilter.php:59]: My motivation here is to make it work with Symfony v4. There are certainly a lot of ways one could improve this.

File: composer.json (hunk 6 above)
Comment (Macavity) [reply composer.json:17]: I thought that as well, but there are conflicts with older versions.

File: .travis.yml (hunk 7 above)
Comment (Macavity) [reply .travis.yml:33]: If it would work I would have kept it.

**Comment by Macavity** [comment]
Sorry, my motivation was to make assetic work with Symfony 4. This PR does just that. There are many comments about potential changes, but personally, I don't get the feeling that the changes will get the PR approved.

By the way, just allowing symfony 4 as a dependency as done in #866 broke when we tried to compile assets.

If you need to get assetic working with Symfony 4 you can just link the package to my repository. That works fine for us.

**Comment by LukeTowers** [comment]
This has been implemented in the 2.0 branch of [assetic-php/assetic](https://github.com/assetic-php/assetic). Until we tag 2.0, you can pull it in by using `"assetic/framework": "dev-2.0/dev"` in your composer.json files.

**Pull request closed without merging**